### ISSUE TITLE
How strictly are constraints enforced?

### ISSUE THREAD

**Issue by mgarort** [cornellius-gp/gpytorch#849]
Hi,

I have found that, while optimizing, sometimes parameters get out of the bounds determined by the constraints that I have specified (even if they started optimization well within these bounds). Is this a bug or the expected behaviour if the optimization is challenging?

For me it happens often with the noise variance. For instance, I may specify

```python
likelihood = gpytorch.likelihoods.GaussianLikelihood(noise_constraint=gpytorch.constraints.LessThan(30.))
```

But I end up with
```
likelihood.noise_covar.raw_noise Parameter containing:
tensor([801.9753], dtype=torch.float64, requires_grad=True)
```

(With such high noise my model is hugely overfitting the training set...)

Thanks!

**Comment by jacobrgardner** [comment]
@mgarort The way the constraints are enforced is by warping the raw values so that `likelihood.noise` is constrained (hence the term `raw`. In your case, even when the `raw_noise` is 801, the true noise will be clipped in a differentiable way to 30:

```python
likelihood = gpytorch.likelihoods.GaussianLikelihood(noise_constraint=gpytorch.constraints.LessThan(30.))
likelihood.noise_covar.raw_noise.fill_(801)
print(likelihood.noise)  # torch.tensor([30.], grad_fn=<AddBackward0>)
```

If I had to hazard a guess, the real problem here is that if you are ending up with noises that large, your training data and labels are almost certainly not normalized. If you normalize them, these problems will probably go away.

**Comment by mgarort** [comment]
I see! Thanks, that's much clearer.

You are right, I'm not normalizing the data. When you say the training data and labels, do you mean the features X and the targets Y respectively? A couple of questions about these:

- I used to always center the targets Y because usually we set a zero-mean prior for them. However, I have noticed that in practice it doesn't make a difference to center them or not before training. Is this because of the constant mean function in GPytorch?

- Is there any reason why I should normalize the features X? If I am using a kernel with automatic relevance determination for the lengthscales, in theory the scale of the features shouldn't matter.

**Comment by jacobrgardner** [comment]
@mgarort Sorry for the delayed reply! We've been pretty busy working on an aistats submission.

Normalization matters because of the hyperparameter initialization. If your features are on the order of thousands, and as a result your distances are on the order of thousands, but your lengthscale is initialized to say 1, then your RBF kernel will try to compute `exp(-1000/1) = exp(-1000)`. This is 0 to numerical precision. As a result, your kernel matrix will be the identity matrix to begin with, and changing the lengthscale by a small amount in either direction won't change this.

On the flip side, if your distances are tiny because your featuers are tiny, then you could get RBF kernel values like `exp(-0.00001/1) = 1`. Then, your kernel matrix will be the matrix of all 1s to machine precision and won't even be invertible.

The same problem can happen with the labels, specifically affecting (1) the noise (obviously the amount of noise changes with the scale of your labels), and (2) the outputscale. This is less problematic, because it doesn't outright lead to the NLL being uncomputable in floating point, but it can make convergence much much slower because your initial hypers will be very far from what they need to be.
